File: src/main/java/mnm/mods/tabbychat/client/gui/component/config/GuiSettingNumber.java
package mnm.mods.tabbychat.client.gui.component.config;

import mnm.mods.tabbychat.util.config.Value;
import mnm.mods.tabbychat.client.gui.component.GuiNumericUpDown;
import mnm.mods.tabbychat.client.gui.component.config.GuiSetting.GuiSettingWrapped;

/**
 * A base gui setting for numbers. It wraps a {@link GuiNumericUpDown}
 *
 * @author Matthew
 * @param <T>
 * @see GuiSettingDouble
 * @see GuiSettingInt
 */
public abstract class GuiSettingNumber<T extends Number> extends GuiSettingWrapped<T, GuiNumericUpDown<T>> {

    private GuiSettingNumber(Value<T> setting, GuiNumericUpDown<T> input) {
        super(setting, input);
    }

    /**
     * Gui setting for integers
     *
     * @author Matthew
     */
    public static class GuiSettingInt extends GuiSettingNumber<Integer> {

        public GuiSettingInt(Value<Integer> setting) {
            super(setting, new GuiNumericUpDown.IntUpDown());
        }
    }

    /**
     * Gui setting for doubles
     *
     * @author Matthew
     */
    public static class GuiSettingDouble extends GuiSettingNumber<Double> {

        public GuiSettingDouble(Value<Double> setting) {
            super(setting, new GuiNumericUpDown.DoubleUpDown());
        }
    }
}

File: src/main/kotlin/mnm/mods/tabbychat/client/gui/component/config/GuiSettingNumber.kt
package mnm.mods.tabbychat.client.gui.component.config

import mnm.mods.tabbychat.client.gui.component.GuiNumericUpDown
import mnm.mods.tabbychat.client.gui.component.config.GuiSettingNumber.GuiSettingDouble
import mnm.mods.tabbychat.client.gui.component.config.GuiSettingNumber.GuiSettingInt
import kotlin.reflect.KMutableProperty0

/**
 * A base gui setting for numbers. It wraps a [GuiNumericUpDown]
 *
 * @param <T>
 * @see GuiSettingDouble
 * @see GuiSettingInt
 */
abstract class GuiSettingNumber<T : Number>
private constructor(setting: KMutableProperty0<T>, input: GuiNumericUpDown<T>)
    : GuiSetting.ValueSetting<T, GuiNumericUpDown<T>>(setting, input) {

    /**
     * Gui setting for integers
     */
    class GuiSettingInt(setting: KMutableProperty0<Int>) : GuiSettingNumber<Int>(setting, GuiNumericUpDown.IntUpDown())

    /**
     * Gui setting for doubles
     */
    class GuiSettingDouble(setting: KMutableProperty0<Double>) : GuiSettingNumber<Double>(setting, GuiNumericUpDown.DoubleUpDown())
}
